### PR TITLE
Graphics refactor #8

### PR DESCRIPTION
**Pull request by kkaefer**

### CODE REVIEW

File: platform/node/src/node_mapbox_gl_native.cpp [hunk 1]
```diff
@@ -6,19 +6,32 @@
...
 #include "node_expression.hpp"
 
+
+void SetBackendType(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+    if (info.Length() < 1 || info[0]->IsUndefined()) {
+        return Nan::ThrowTypeError("Requires a render backend name");
+    }
+
+    const std::string backendName { *Nan::Utf8String(info[0]) };
+    (void)backendName;
```
Comment (pozdnyakov) [inline platform/node/src/node_mapbox_gl_native.cpp]: could you explain why this is needed?

File: include/mbgl/gfx/backend.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    enum class Type : uint8_t {
+        OpenGL,
+    };
+
+    static constexpr Type DefaultType = Type::OpenGL;
+
+    static void SetType(const Type value) {
+        if (Value(value) != value) {
+            abort(); // SetType must be called prior to any GetType calls.
+        }
```
Comment (pozdnyakov) [inline include/mbgl/gfx/backend.hpp]:
```
Value(value);
assert(Value(value) == value);
```
?

File: src/mbgl/renderer/renderer_impl.cpp [hunk 3]
```diff
@@ -503,6 +503,17 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
...
 
+    // Ends the RenderPass
+    parameters.renderPass.reset();
+
+    if (updateParameters.mode == MapMode::Continuous) {
+        parameters.encoder->present(parameters.backend.getDefaultRenderable());
+    }
+
+    // CommandEncoder destructor submits render commands.
+    parameters.encoder.reset();
```
Comment (pozdnyakov) [inline src/mbgl/renderer/renderer_impl.cpp]: could we instead put `parameters` into a separate scope (or a separate method)?

File: src/mbgl/renderer/renderer_impl.cpp [hunk 4]
```diff
@@ -503,6 +503,17 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
     }
 #endif
 
+    // Ends the RenderPass
+    parameters.renderPass.reset();
```
Comment (pozdnyakov) [inline src/mbgl/renderer/renderer_impl.cpp]: what is the reason it cannot die after `parameters.encoder->present(parameters.backend.getDefaultRenderable());`?

File: src/mbgl/renderer/renderer_impl.cpp (hunk 3 above)
Comment (kkaefer) [reply src/mbgl/renderer/renderer_impl.cpp]: We could, but that's out of scope for this PR; it only moves the code.

File: src/mbgl/renderer/renderer_impl.cpp (hunk 4 above)
Comment (astojilj) [reply src/mbgl/renderer/renderer_impl.cpp]: Clarified offline, it is about order.

File: include/mbgl/gfx/backend.hpp (hunk 2 above)
Comment (kkaefer) [reply include/mbgl/gfx/backend.hpp]: there's no way out if this assert is false, so we can just as well run `abort`

**Pull request merged**